modules/renter/repairscanner.go

```diff
@@ -0,0 +1,333 @@
+	// of memory available, and then spin up a thread to asynchronously handle
+	// the rest of the chunk tasks.
+	memoryAvailable := r.managedMemoryAvailableGet()
+	nextChunk := ch.Pop().(*unfinishedChunk)
+	for nextChunk.memoryNeeded > memoryAvailable {
+		select {
+		case newFile := <-r.newUploads:
+			r.managedInsertFileIntoChunkHeap(newFile, ch, hosts, fcidToHPK)
+		case <-r.newMemory:
+		case <-r.tg.StopChan():
+		}
+	}
+	r.managedMemoryAvailableSub(nextChunk.memoryNeeded)
+	go r.managedFetchAndRepairChunk(nextChunk)
+}
+
+// threadedRepairScan is a background thread that checks on the health of files,
+// tracking the least healthy files and queuing the worst ones for repair.
+//
+// TODO / NOTE: Once we have upgraded the filesystem, we can replace this with
+// the tree-diving technique discussed in Sprint 5. For now we just iterate
+// through all of our in-memory files and chunks, and maintain a finite list of
+// the worst ones, and then iterate through it again when we need to find more
+// things to repair.
+func (r *Renter) threadedRepairScan() {
+	err := r.tg.Add()
+	if err != nil {
+		return
+	}
+	defer r.tg.Done()
+
+	for {
+		// Return if the renter has shut down.
+		select{
+		case <-r.tg.StopChan():
+			return
+		default:
+		}
+
+		// Grab the current set of contracts and a lookup table from file
+		// contract ids to host public keys. The lookup table has a mapping from
+		// all historic file contract ids, which is necessary when using the
+		// renter to perform downloads.
+		//
+		// TODO / NOTE: This code can be removed once files store the HostPubKey
+		// of the hosts they are using, instead of just the FileContractID.
+		currentContracts, fcidToHPK := r.managedCurrentContractsAndHistoricFCIDLookup()
+
+		// Pull together a list of hosts that are available for uploading. We
+		// assemble them into a map where the key is the String() representation
+		// of a types.SiaPublicKey (which cannot be used as a map key itself).
+		hosts := make(map[string]struct{})
+		for _, contract := range currentContracts {
+			hosts[contract.HostPublicKey.String()] = struct{}{}
+		}
+
+		// Build a min-heap of chunks organized by upload progress.
+		chunkHeap := r.managedBuildChunkHeap(hosts, fcidToHPK)
+
+		// Refresh the worker pool before beginning uploads.
+		r.managedUpdateWorkerPool()
+
+		// Work through the heap. Chunks will be processed one at a time until
+		// the heap is whittled down. When the heap is empty, we wait for new
+		// files in a loop and then process those. When the rebuild signal is
+		// received, we start over with the outer loop that rebuilds the heap
+		// and re-checks the health of all the files.
+		rebuildHeapSignal := time.After(rebuildChunkHeapInterval)
+		for {
+			// Return if the renter has shut down.
+			select{
+			case <-r.tg.StopChan():
+				return
+			default:
+			}
+
+			if chunkHeap.Len() > 0 {
+				r.managedPrepareNextChunk(chunkHeap, hosts, fcidToHPK)
+			} else {
```

Comment:
I prefer to avoid large `else` blocks:
```go
if chunkHeap.Len() > 0 {
    r.managedPrepareNextChunk(chunkHeap, hosts, fcidToHPK)
    continue
}
// rest of code
```